### PR TITLE
#99 - Touch ID Preprocessing

### PR DESCRIPTION
**Pull request by Russell-Newton**
Touch IDs are assigned according to the following rules:
* Automatic assignment only happens when the `touch_event.id` is `TOUCH_ID_UNDEFINED`.
* A set of "group heads" the size of `MAX_TOUCHES` stores the latest touch assigned to each group.
  * Groups are considered to be untracked if their head is a `TOUCH_EVENT_UP` touch.
  * All groups are initialized to being untracked with `init_gesturelib()`.
* A `TOUCH_EVENT_DOWN` touch will be assigned to the first group with a previous `TOUCH_EVENT_UP` head.
  * If no group is free, don't update `touch_event.id`, return `TOUCH_ID_UNDEFINED`.
* A `TOUCH_EVENT_MOVE` or `TOUCH_EVENT_UP` touch will be assigned to the group with a head closest to this touch's position.
  * If no groups are being tracked, don't update `touch_event.id`, return `TOUCH_ID_UNDEFINED`.
  * Head proximity is calculated with Euclidean.

* When groups are set to be untracked, other groups are not affected. The third finger placed will always have touches in the third group, even if the first and second fingers are lifted up.
  * This is important for correctly recognizing gestures.

### CODE REVIEW

File: gesturelibrary/include/gesturelib.h [hunk 1]
```diff
@@ -59,6 +62,15 @@ void init_gesturelib();
...
 int process_touch_event(touch_event_t* touch_event, gesture_event_t* gestures, int max_gestures);
 
+/// @brief assign a group ID to this touch_event. Update group_heads to reflect this. If this is not a DOWN event
+///        and there are no groups being tracked, do not assign an ID. If this is a DOWN event but MAX_TOUCHES groups
+///        are already being tracked, do not assign an ID.
+/// @param touch_event
+/// @return
+unsigned int assign_group(touch_event_t* touch_event);
+
+float squared_distance(touch_event_t* a, touch_event_t* b);
```
Comment (xorbit3024) [inline gesturelibrary/include/gesturelib.h:72]: I don't think this makes sense to expose as part of the gesturelib API.
Please declare it static in gesturelib.c

File: gesturelibrary/include/gesturelib.h [hunk 2]
```diff
@@ -78,7 +90,3 @@ int enable_recognizer(int recognizer);
 /// @param recognizer which recognizer to disable
 /// @return 0 if an invalid index is provided, 1 otherwise
 int disable_recognizer(int recognizer);
-
-/// @brief
-/// @return the number of registered recognizers
-int get_num_recognizers();
```
Comment (xorbit3024) [inline gesturelibrary/include/gesturelib.h:84]: Why was this removed?

File: gesturelibrary/include/gesturelib.h (hunk 1 above)
Comment (xorbit3024) [reply gesturelibrary/include/gesturelib.h:72]: Alternatively, make a utils folder and place it there.

File: gesturelibrary/include/gesturelib.h (hunk 2 above)
Comment (Russell-Newton) [reply gesturelibrary/include/gesturelib.h:84]: It shouldn't be removed. I'll look at that again when I get back home

File: gesturelibrary/include/gesturelib.h (hunk 1 above)
Comment (Russell-Newton) [reply gesturelibrary/include/gesturelib.h:72]: that makes sense. I'll make it private in the c file

File: gesturelibrary/include/gesturelib.h (hunk 2 above)
Comment (Russell-Newton) [reply gesturelibrary/include/gesturelib.h:84]: Oh I remember now. It's because I exposed the `num_recognizers` variable as extern within the header file. Didn't need to have a getter for it.

File: gesturelibrary/src/gesturelib.c [hunk 3]
```diff
@@ -9,11 +9,21 @@
 gesture_recognizer_t recognizers[MAX_RECOGNIZERS];
 int num_recognizers = 0;
 
+/// @brief set containing most recent touch within finger group
+touch_event_t* groups_heads[MAX_TOUCHES];
+touch_event_t empty_touch_event = {
+    .event_type = TOUCH_EVENT_UP, .position_x = 0, .position_y = 0, .timestamp = 0, .id = TOUCH_ID_UNDEFINED};
+
+float squared_distance(touch_event_t* a, touch_event_t* b);
```
Comment (xorbit3024) [inline gesturelibrary/src/gesturelib.c:17]:
```suggestion
static float squared_distance(touch_event_t* a, touch_event_t* b);
```

File: gesturelibrary/src/gesturelib.c [hunk 4]
```diff
@@ -35,6 +48,60 @@ int process_touch_event(touch_event_t* touch_event, gesture_event_t* gestures, i
...
+        return TOUCH_ID_UNDEFINED;
+    }
+
+    touch_event->id             = closest_group;
+    groups_heads[closest_group] = touch_event;
+
+    return closest_group;
+}
+
+float squared_distance(touch_event_t* a, touch_event_t* b) {
```
Comment (xorbit3024) [inline gesturelibrary/src/gesturelib.c:99]:
```suggestion
static float squared_distance(touch_event_t* a, touch_event_t* b) {
```

**Review by xorbit3024 — APPROVED** [review]
